packages/client/hmi-client/src/components/responsive-matrix/label-cols.vue

```diff
@@ -0,0 +1,173 @@
+	// props                                                                        //
+	// ---------------------------------------------------------------------------- //
+
+	props: {
+		viewport: {
+			type: null,
+			default() {
+				return null;
+			}
+		},
+		numCols: {
+			type: Number,
+			default() {
+				return 0;
+			}
+		},
+		update: {
+			type: Number,
+			default() {
+				return 0;
+			}
+		},
+		move: {
+			type: Number,
+			default() {
+				return 0;
+			}
+		},
+		selectedCols: {
+			type: Array as PropType<CellStatus[]>,
+			default() {
+				return [];
+			}
+		},
+		labelColList: {
+			type: Array as PropType<number[] | string[]>,
+			default() {
+				return [];
+			}
+		},
+		microColSettings: {
+			type: Array as PropType<number[]>,
+			default() {
+				return [];
+			}
+		}
+	},
+
+	// ---------------------------------------------------------------------------- //
+	// data                                                                         //
+	// ---------------------------------------------------------------------------- //
+
+	data() {
+		return {
+			labelStyle: {},
+			labelPositions: [] as number[],
+			labelValues: [] as any[]
+		};
+	},
+
+	// ---------------------------------------------------------------------------- //
+	// mounted                                                                      //
+	// ---------------------------------------------------------------------------- //
+
+	mounted() {
+		this.getColLabelStyle();
+		this.buildLabelData();
+	},
+
+	// ---------------------------------------------------------------------------- //
+	// watch                                                                        //
+	// ---------------------------------------------------------------------------- //
+
+	watch: {
+		move() {
+			this.getColLabelStyle();
+		},
+		update() {
+			this.buildLabelData();
+		}
+	},
+
+	// ---------------------------------------------------------------------------- //
+	// methods                                                                      //
+	// ---------------------------------------------------------------------------- //
+
+	methods: {
+		buildLabelData() {
+			const visibleBounds = this.viewport?.getVisibleBounds();
+			const viewportColDensity =
+				((visibleBounds?.width || 0) / (this.viewport?.worldWidth || 1)) * this.numCols;
+
+			const thresholdLabelDensity = 8;
+			const labelStride = Math.max(1, Math.ceil(viewportColDensity / thresholdLabelDensity));
+
+			this.labelPositions = [];
+			this.labelValues = [];
+			let pos = 0;
+			for (let i = 0; i < this.selectedCols.length; i++) {
+				const len = this.microColSettings[this.selectedCols[i]];
+
+				if (!(i % labelStride)) {
+					this.labelPositions.push(pos + len / 2);
+					this.labelValues.push(this.labelColList[i]);
+				}
+
+				pos += len;
+			}
+
+			this.labelPositions = this.labelPositions.map((v) => (v / pos) * 100);
+		},
```

Comment:
This can be factored out and shared I think , there shouldn't be much difference in terms of the calculation no?
Also a little bit of comment would help

Reply:
ive considered factoring this out, but didnt because i thought it would be messy having to add args so that i can switch around col/row and width/height everywhere that thats required. thoughts?

Reply:
Sorry I didn't necessarily mean the entire function, but how things are cut up and proportioned - I will take a crack at this.